crates/biome_js_semantic/src/format_semantic_model.rs

```diff
@@ -0,0 +1,201 @@
+            print_width: self.line_width().into(),
+            line_ending: self.line_ending(),
+            indent_style: self.indent_style(),
+        }
+    }
+}
+
+struct FormatSemanticModelContext;
+
+impl FormatContext for FormatSemanticModelContext {
+    type Options = FormatSemanticModelOptions;
+
+    fn options(&self) -> &Self::Options {
+        &FormatSemanticModelOptions
+    }
+
+    fn source_map(&self) -> Option<&TransformSourceMap> {
+        None
+    }
+}
+
+impl std::fmt::Display for SemanticModel {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let formatted = biome_formatter::format!(FormatSemanticModelContext, [&self])
+            .expect("Formatting not to throw any FormatErrors");
+        f.write_str(
+            formatted
+                .print()
+                .expect("Expected a valid document")
+                .as_code(),
+        )
+    }
+}
+
+impl Format<FormatSemanticModelContext> for SemanticModel {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        write!(f, [self.global_scope()])
+    }
+}
+
+enum ScopeOrBinding {
+    Scope(Scope),
+    Binding(Binding),
+}
+
+impl Format<FormatSemanticModelContext> for Scope {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        let mut children = self
+            .descendents()
+            .skip(1) // skips the current scope
+            .map(ScopeOrBinding::Scope)
+            .chain(self.bindings().map(ScopeOrBinding::Binding))
+            .collect::<Vec<_>>();
+        // Yes it would be more efficient to just build the vector in the right order, but
+        // this is easier to read and maintain.
+        children.sort_by_key(|item| match item {
+            ScopeOrBinding::Scope(scope) => scope.range().start(),
+            ScopeOrBinding::Binding(binding) => binding.syntax().text_range_with_trivia().start(),
+        });
+
+        write!(
+            f,
+            [
+                text("Scope {"),
+                group(&block_indent(&format_args![
+                    format_with(|f| {
+                        let range = std::format!("{:?}", self.range());
+                        write!(
+                            f,
+                            [
+                                text("id: "),
+                                self.id,
+                                text(" @ "),
+                                dynamic_text(range.as_str(), TextSize::default()),
+                                hard_line_break()
+                            ]
+                        )?;
+
+                        Ok(())
+                    }),
+                    text("children ["),
+                    group(&block_indent(&format_with(|f| {
+                        if children.is_empty() {
+                            return write!(f, [text("No children")]);
+                        }
+                        for child in &children {
```

Comment:
A list should be printed using the joiner utility. One API is `f.join()`, but there are others